### PR TITLE
Initial kinda working sync over the network!

### PR DESCRIPTION
**Pull request by rictic**
So, the borrow checker was indeed unhappy with there being two
flows where changes could come in – the network and the main JS loop.

The solution was to use a wrapper to synchronize, an Arc<Mutex<Game>>.
I think this is a fairly common pattern for dealing with mutation
from multiple sources.

Arc<X> is a reference counted wrapper around X. It's like Box<X> only
there can be more than one copy of the pointer to the value, and
the data will stick around as long as at least one pointer lives. Docs here: https://doc.rust-lang.org/std/sync/struct.Arc.html

Mutex<X> defines a read/write lock around X, such that you can only
read or write the underlying X value while you have a lock on it.
When you're done with the value you automatically release the lock.
Pretty handy! I've always seen it used with Arc, I guess because if
there's only one copy of the Mutex then there's no point in the Mutex
because ordinary Rust lifetimes give you the guarantees of a lock.
Docs here: https://doc.rust-lang.org/std/sync/struct.Mutex.html

I _think_ that we'll never be in a situation where the lock is held
by multiple things at the same time, because of the guarantees of the
browser's event loop – we're fundamentally single threaded at this
point. However Rust can't know that, and in the future we can use
SharedArrayBuffer anyways, so this seems like good discipline.

Also I don't know how to opt out anyways, so it's not like we have any
other options lol.

So with this in mind, we don't pass back a real Game to the JS, just
a wrapper around an Arc<Mutex<Game>>.

In a follow up change I'll clean up the simulate / simulate_internal now
that Game is internal.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -384,67 +439,86 @@ pub fn make_game() -> Result<Game, JsValue> {
...
-        players,
-        socket: ws,
+    let wrapper = GameWrapper {
+        game: Arc::new(Mutex::new(Game {
+            speed: 2.0,
+            context,
+            width,
+            height,
+            kill_distance: 64.0,
+            local_player_color: Some(Color::random()),
```
Comment (cdata) [inline src/lib.rs:449]: Ooo `Color::random()` :gem:

File: src/lib.rs [hunk 2]
```diff
@@ -384,67 +439,86 @@ pub fn make_game() -> Result<Game, JsValue> {
...
-        speed: 2.0,
-        context,
-        width,
-        height,
-        kill_distance: 64.0,
-        local_player_color: Some(Color::random()),
-        players,
-        socket: ws,
+    let wrapper = GameWrapper {
+        game: Arc::new(Mutex::new(Game {
```
Comment (cdata) [inline src/lib.rs:443]: is `::new` a language-level concept or just a conventional name for part of the implementation?

File: src/lib.rs [hunk 3]
```diff
@@ -384,67 +439,86 @@ pub fn make_game() -> Result<Game, JsValue> {
...
-            let message: Message = match serde_json::from_str(&strng) {
-                Ok(m) => m,
-                Err(e) => {
-                    console_log!("Unable to deserialize {:?} – {:?}", strng, e);
-                    return;
+    {
+        let game = wrapper
+            .game
+            .lock()
+            .expect("Could not get a lock on the game we just created??");
```
Comment (cdata) [inline src/lib.rs:459]: Is it because of an fold in time and space!!

File: src/lib.rs [hunk 4]
```diff
@@ -384,67 +439,86 @@ pub fn make_game() -> Result<Game, JsValue> {
...
+
+        let game_clone = wrapper.game.clone();
+        let onmessage_callback = Closure::wrap(Box::new(move |e: MessageEvent| {
+            // Starting with assuming text messages. Can make efficient later (bson?).
```
Comment (cdata) [inline src/lib.rs:463]: For a game like this it may not make that much of a difference, but you could get very compact with your protocol for most things

**Review by cdata — APPROVED** [review]
Thanks for the great high-level explanation of Arc and Mutex!

File: src/lib.rs (hunk 2 above)
Comment (rictic) [reply src/lib.rs:443]: The latter, a conventional name for an otherwise totally normal function

File: src/lib.rs (hunk 3 above)
Comment (rictic) [reply src/lib.rs:459]: All bets: off

File: src/lib.rs (hunk 4 above)
Comment (rictic) [reply src/lib.rs:463]: Yeah!

What's the size of packets you can count on? 1.5KB? That's a _huge_ amount of headspace at this point! Who could ever fill up that much space!?

**Pull request merged**